### ISSUE TITLE
[Bug] StackOverflowException in UWP ListViewRenderer

### ISSUE THREAD

**Issue by SuNNjek** [xamarin/Xamarin.Forms#8278]
### Description
When using a `List<KeyValuePair<string, string>>` as the ItemsSource for a ListView, the UWP ListViewRenderer crashes with a StackOverflowException when clicking on an item. The same code works on Android and iOS.

### Steps to Reproduce

1. Run the test application attached below on UWP
2. Click on any item in the ListView
3. The application crashes with a StackOverflowException

### Expected Behavior
On tapping the item, an alert should show displaying the selected item.

### Actual Behavior
The app crashes

### Basic Information

- Version with issue: 4.2.0.709249
- Last known good version: -
- IDE: Visual Studio 2019 Professional
- Platform Target Frameworks: 
  - iOS:  13
  - Android: 9.0
  - UWP:  16299
- Android Support Library Version: - 
- Nuget Packages: -
- Affected Devices: Windows PC

### Reproduction Link

[ListViewBug.zip](https://github.com/xamarin/Xamarin.Forms/files/3782448/ListViewBug.zip)

**Comment by WayaFlyfeather** [comment]
I can confirm this happens.

The exception is thrown from [here in the UWP ListViewRenderer](https://github.com/WayaFlyfeather/Xamarin.Forms/blob/d4cb7ec0b09ae0f56aad6ae8a260f278465146a3/Xamarin.Forms.Platform.UAP/ListViewRenderer.cs#L780)

It appears that the List.SelectedItem and Element.SelectedItem are not equal, even though the two KeyValuePair<string,string> appears equal in the debugger. I assume this is caused by them being boxed on different occasions(?)

I experimented with breaking the recursion by setting _itemWasClicked=false just before setting List.SelectedItem, like this:

```
if (_itemWasClicked)
{
    _itemWasClicked = false;
    List.SelectedItem = Element.SelectedItem;
}
```

This seems to work, but I guess it is probably not the correct approach - it seems to me that there may be a deeper problem with boxing and comparison. (I also tried to make the test case with a list of ints, to see if it happened to all Value Types, but this seems to work - no recursion.)
